llvm/lib/Analysis/LoopAccessAnalysis.cpp

```diff
@@ -1039,13 +1040,10 @@ findForkedPointer(PredicatedScalarEvolution &PSE,
   SmallVector<PointerIntPair<const SCEV *, 1, bool>> Scevs;
   findForkedSCEVs(SE, L, Ptr, Scevs, MaxForkedSCEVDepth);
 
-  // For now, we will only accept a forked pointer with two possible SCEVs
-  // that are either SCEVAddRecExprs or loop invariant.
+  // For now, we will only accept a forked pointer with two possible SCEVs.
   if (Scevs.size() == 2 &&
-      (isa<SCEVAddRecExpr>(get<0>(Scevs[0])) ||
-       SE->isLoopInvariant(get<0>(Scevs[0]), L)) &&
-      (isa<SCEVAddRecExpr>(get<0>(Scevs[1])) ||
-       SE->isLoopInvariant(get<0>(Scevs[1]), L))) {
+      hasComputableBounds(PSE, Ptr, get<0>(Scevs[0]), L, /*Assume*/ false) &&
+      hasComputableBounds(PSE, Ptr, get<0>(Scevs[1]), L, /*Assume*/ false)) {
```

Comment:
It is not immediately clear to me how this is NFC in all cases, as `hasComputableBounds` also checks if the SCEV is affine?

It makes sense to have this restriction for forked pointers as well, but it might need a test?

Reply:
The only user of findForkedPointer function is the AccessAnalysis::createCheckForAccess method, and it will check later that each translated pointer has computable bounds:https://github.com/llvm/llvm-project/blob/3ec1b1a4dd52641c4b84cac55ef3a228960a0bdc/llvm/lib/Analysis/LoopAccessAnalysis.cpp#L1072 So here we just move this check earlier. I can imagine it can work slightly differently in the following situation: we have a forked pointer and some traslated pointer of it is non-affine AddRec. In this case, the current implementation will proceed it as 2 forked pointers and will bail out on hasComputableBounds check; the new implementation will detect this earlier and findForkedPointer will return the single non-translated pointer. I assume there will be no difference because this pointer also should fail hasComputableBounds check (if one of the translated pointers was a non-affine AddRec, it's hard to imagine that the original pointer can be affine).

BTW, getStartAndEndForAccess method will not handle non-affine case correctly anyway: it takes start and end values from the first/last loop iteration, and for non-linear SCEVs the bound can be reached at some other iteration.

Reply:
Gentle ping